Payouts standard (#192)
near/NEPs: specs/Standards/NonFungibleToken/payouts.md

# Standard for a Multiple-Recipient-Payout mechanic on NFT Contracts

> A standard for supporting structures that allow for NFT's to inform financial

Reviewer: ```suggestion
A standard for supporting structures that allow for NFTs to inform financial
```

Apostrophe only needed here to indicate possession. We wouldn't write "Non-Fungible Token's," so we don't write "NFT's."

Change needed thoughout document.

> contracts to payout multiple addresses on sales.

Reviewer: ```suggestion
contracts to pay out multiple addresses on sales.
```

"Payout" is a noun; "pay out" is the verb. (Similar to "What's your login?" vs "Please log in")

> Currently, NFT's on Near support the field `owner_id`, but lack flexibility for

Reviewer: This is written from the perspective of the present, but is meant to be a reference document for the future. Should it be worded to tell people why they would want to use this NFT extension, rather than telling the repository maintainers why they should merge it?

ownership and payout mechanics with more complexity, including but not limited
to royalties. Financial contracts, such as Marketplaces, Auction-houses, and
NFT Loan contracts, would benefit from a standard interface on NFT producer

> contracts for querying whom to pay out, and how much to pay.

Reviewer: ```suggestion
contracts for querying who to pay and how much.
```

Therefore, the core goal of this standard is to define a set of methods for

> financial standard set of methods to call, without specifying how NFT contracts

Reviewer: ```suggestion
financial contracts to call, without specifying how NFT contracts
```

I think this is what you meant?

define the divide of payout mechanics.

> The `Payout` interface is should be a simple collection of unique addresses that

Reviewer: ```suggestion
The `Payout` interface is a simple collection of unique addresses that
```

map to either a proportional ownership of a Token, or even simpler, a mapping of
unique addresses to the Balances each is to receive, given an initial balance:
```rust
    // return fractional ownership
    pub struct OwnershipFractions {

>       pub fractions: HashMap<AccountId, MultipliedSafeFraction>

Reviewer: I like your suggestion to go with the latter. I think in a proposal of a new standard, you could just propose your best guess in the main PR, and offer alternatives as PRs targeting the main PR. Or in comments, if a second PR seems like overkill or is hard to pull off for non-maintainers of the repo.

    }
    // or else, return how much to pay each address
    pub struct Payout {
      pub payout: HashMap<AccountId, U128>,
    }
```
The advantage of the latter interface eliminates the need for a universally
respected fraction type, instead simply returning the amount to pay out to each
address.

## Methods

To support the payout interface, several methods should be supported on NFT
contracts:

-   A view method to query whom would be paid out for a given token, derived from
    some balance
-   A call method to transfer attempt to transfer the token, which returns the payout
    data

The latter of these two methods should be functionally equivalent to the
composition of `nft_transfer` in the Core standard, while returning the result
of the first method.
```rust
    pub trait Payouts{
      fn nft_payout(&self, token_id: String, balance: U128) -> Payout;

>       fn nft_transfer_payout(

Reviewer: I saw @mattlockyer implement this interface as a modification to the core `nft_transfer` method, since `balance` argument is optional and returned `Payout` is also optional. This keeps the interface minimal and keeps backwards compatibility with current standard. I think I like that as a starting point for the discussion.

Reviewer: Matt's implementation here https://github.com/near-apps/nft-market/blob/387510135f12a3ee4e7980db4b69fbbb9f8ca8da/contracts/nft-simple/src/nft_core.rs#L97

Reviewer 2: Let's not change the return value on a core method. Payouts are very specific to a subset of NFTs using marketplaces and royalties. Let's stick to having it a separate method.
